bin/build-drivers/build_drivers/checksum.clj

```diff
@@ -0,0 +1,76 @@
+(ns build-drivers.checksum
+  "Shared code for calculating and reading hex-encoded MD5 checksums for relevant files."
+  (:require [build-drivers
+             [common :as c]
+             [plugin-manifest :as manifest]]
+            [clojure.java.io :as io]
+            [clojure.string :as str]
+            [metabuild-common.core :as u])
+  (:import org.apache.commons.codec.digest.DigestUtils))
+
+(defn checksum-from-file
+  "Read a saved MD5 hash checksum from a file."
+  [filename]
+  (u/step (format "Read saved checksum from %s" filename)
+    (let [file (io/file filename)]
+      (if-not (.exists file)
+        (u/announce "%s does not exist" filename)
+        (when-let [[checksum-line] (not-empty (str/split-lines (slurp file)))]
+          (when-let [[_ checksum-hex] (re-matches #"(^[0-9a-f]{32}).*$" checksum-line)]
+            (u/safe-println (format "Saved checksum is %s" checksum-hex))
+            checksum-hex))))))
+
+;;; -------------------------------------------- Metabase source checksum --------------------------------------------
+
+(defn- metabase-source-paths []
+  (sort
+   (cons
+    (c/filename c/project-root-directory "project.clj")
+    (mapcat (fn [dir]
+              (try
+                (u/find-files dir #(str/ends-with? % ".clj"))
+                (catch Throwable _
+                  [])))
+            [(c/filename c/project-root-directory "src")
+             (c/filename c/project-root-directory "enterprise" "backend" "src")
+             (c/filename c/project-root-directory "backend" "mbql")]))))
+
+(defn metabase-source-checksum
+  "Checksum of Metabase backend source files and `project.clj`."
+  ^String []
+  (let [paths (metabase-source-paths)]
+    (u/step (format "Calculate checksum for %d Metabase source files" (count paths))
+      (let [checksum (DigestUtils/md5Hex (str/join (map slurp paths)))]
```

Comment:
Shouldn't we start getting into the habit of avoiding MD5? I realize this is a very low risk use, though.

Reply:
It's just for determining whether we need to rebuild stuff or not, not for any sort of secure checksum